fix: added pagination in explore page

Anshumancanrock opened the pull request:
### Description

Added pagination functionality to the Object Explorer page (`/resources`) to improve performance and user experience when viewing large numbers of Kubernetes resources. Previously, all resources were displayed on a single page, which could lead to performance issues and poor UX with hundreds of objects. Now, resources are paginated with 10 items per page, making it easier to browse and navigate through resources.

The pagination implementation follows the existing design patterns used in the codebase (specifically `BPPagination.tsx`) and includes smooth navigation controls with proper styling for both light and dark themes.

### Related Issue

Fixes #2124 

### Video after fix:


https://github.com/user-attachments/assets/45de60fb-41a5-485e-9dd1-6b1c049328f9



### Changes Made

- [x] Added pagination state management (`currentPage`, `itemsPerPage`) to ObjectFilterPage
- [x] Implemented pagination calculation logic using `useMemo` for performance optimization
- [x] Created reusable `renderPagination()` function to avoid code duplication across view modes
- [x] Added auto-reset functionality that resets to page 1 when filters change
- [x] Integrated pagination UI in all three view modes (Grid, List, and Table views)
- [x] Applied consistent styling matching the existing `BPPagination` component design
- [x] Used existing i18n translation keys (`common.previous`, `common.next`, `common.page`, etc.)

### Checklist

Please ensure the following before submitting your PR:

- [x] I have reviewed the project's contribution guidelines.
- [x] I have written unit tests for the changes (if applicable).
- [x] I have updated the documentation (if applicable).
- [x] I have tested the changes locally and ensured they work as expected.
- [x] My code follows the project's coding standards.

Anshumancanrock committed:
fix: added pagination in explore page

alokdangre commented:
@Anshumancanrock fix formatting issue

alokdangre commented:
by the way in video why there is a gap at the end , you can paginate like 12 per page or 9 per page for covering that gap

Anshumancanrock committed:
fix: change items per page from 10 to 9 to avoid layout gaps

Anshumancanrock commented:
Hi @alokdangre , I have fixed the formatting issues and changed items per page from "10" to "9" as you requested. Please review now and let me know if you want any other changes!

alokdangre commented:
/ok-to-test

alokdangre commented:
/lgtm

alokdangre commented:
/cc @btwshivam

alokdangre commented:
@Anshumancanrock  you can update the video with the video after making changes

Anshumancanrock commented:
> @Anshumancanrock you can update the video with the video after making changes

Sure @alokdangre !

Here is the video after latest fix:



https://github.com/user-attachments/assets/1c01e574-a4fa-458c-a17c-33617afec20e

btwshivam commented:
@Anshumancanrock After I approve your PR, please also create a follow-up issue for refactoring. The Object Explorer file has now grown to around 1800+ lines, but we want to keep it within a maximum of 800 lines as part of our ongoing optimization efforts. Once the issue is created, let’s move this into a separate directory and break the file into smaller, modular components for better structure. Also, while doing this, make sure to move the smaller existing components like Preview and others into the same directory so everything stays organized.

btwshivam commented:
@Anshumancanrock But for this PR also make pagination a component and reuse

btwshivam commented on frontend/src/pages/ObjectFilterPage.tsx:
```diff
@@ -378,6 +380,19 @@ const ObjectFilterPage: React.FC = () => {
   }, [derivedResources]);
 
+  // Pagination calculations
+  const totalPages = Math.ceil(derivedResources.length / itemsPerPage);
```
always use usememo

btwshivam commented on frontend/src/pages/ObjectFilterPage.tsx:
```diff
@@ -524,6 +539,161 @@ const ObjectFilterPage: React.FC = () => {
 
+  // Reusable pagination component
+  const renderPagination = () => {
+    if (derivedResources.length === 0) return null;
```
Yeah correct make changes

Anshumancanrock committed:
Optimize with useMemo for performance

Anshumancanrock committed:
fix formatting

Anshumancanrock commented:
@btwshivam I have created a separate pagination component as you requested and am reusing it. I also applied Copilot’s suggestion. I think it’s now fully optimized. Please take a look. Thanks!

Anshumancanrock commented:
> @Anshumancanrock After I approve your PR, please also create a follow-up issue for refactoring. The Object Explorer file has now grown to around 1800+ lines, but we want to keep it within a maximum of 800 lines as part of our ongoing optimization efforts. Once the issue is created, let’s move this into a separate directory and break the file into smaller, modular components for better structure. Also, while doing this, make sure to move the smaller existing components like Preview and others into the same directory so everything stays organized.

Yeah sure! I’ll create the follow-up issue right after the PR is approved and handle the refactor + reorganization as suggested.

onkar717 commented:
@btwshivam ?

btwshivam commented:
Remove hardcoded string and update final vid

Anshumancanrock committed:
fix

Anshumancanrock committed:
fix

Anshumancanrock committed:
Merge branch 'dev' into pagination

Anshumancanrock commented:
done @btwshivam

Anshumancanrock committed:
Merge branch 'dev' into pagination

Anshumancanrock commented:
@btwshivam checks are successful now. Could you please review it finally ? Thanks

btwshivam commented:
/approve
/lgtm
/ok-to-test
